File: 프로그래머스/lv0/120866. 안전지대/안전지대.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<vector<int>> board)
{
    int answer = 0;
    int n = board.size();
    int X[8] = {1, 1, 1, 0, 0, -1, -1, -1};
    int Y[8] = {1, 0, -1, 1, -1, 1, 0, -1};

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (board[i][j] == 1)
            {
                for (int k = 0; k < 8; k++)
                {
                    if (i + X[k] >= 0 && i + X[k] < n && j + Y[k] >= 0 && j + Y[k] < n)
                        if (board[i + X[k]][j + Y[k]] != 1)
                            board[i + X[k]][j + Y[k]] = 2;
                }
            }
        }
    }

    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            if (board[i][j] == 0)
                answer++;

    return answer;
}

File: 프로그래머스/0/120866. 안전지대/안전지대.java
import java.util.*;

class Solution {
    
    static int[] dx = {0, -1, -1, -1, 0, 1, 1, 1, 0};
    static int[] dy = {0, -1, 0, 1, 1, 1, 0, -1, -1};
    
    public int solution(int[][] board) {
        int answer = 0;
        
        int n = board.length, m = board[0].length;
        boolean[][] notSafe = new boolean[n][m];
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 0) continue;
                
                for (int dir = 0; dir < 9; dir++) {
                    int nx = i + dx[dir];    
                    int ny = j + dy[dir];
                    
                    if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
                    notSafe[nx][ny] = true;
                }
            }
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (!notSafe[i][j]) answer++;
            }
        }
        
        
        return answer;
    }
}